pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java

```diff
@@ -643,6 +654,59 @@ public void initializeDispatchRateLimiterIfNeeded(Optional<Policies> policies) {
             this.dispatchRateLimiter = Optional.of(new DispatchRateLimiter(topic, name));
         }
     }
-    
+
+    @Override
+    public synchronized boolean trackDelayedDelivery(long ledgerId, long entryId, MessageMetadata msgMetadata) {
+        if (!isDelayedDeliveryEnabled) {
+            // If broker has the feature disabled, always deliver messages immediately
+            return false;
+        }
+
+        if (!delayedDeliveryTracker.isPresent()) {
+            // Initialize the tracker the first time we need to use it
+            delayedDeliveryTracker = Optional.of(new DelayedDeliveryTracker(this));
+        }
+
+        return delayedDeliveryTracker.get().addMessage(ledgerId, entryId, msgMetadata.getDeliverAtTime());
+    }
+
+    /**
+     * Returns whether we have any message that could be immediately replayed.
+     * This could be a message that was requested to be re-delivered or a delayed
+     * delivery.
+     */
+    private boolean hasMessagesToReplay() {
```

Comment:
I think this method should be synchronized too, because of `delayedDeliveryTracker` and `messagesToRedeliver`

Reply:
Yes, this is only called while already holding the lock. I'll add comment to clarify,

Reply:
Actually, I'll just convert into synchronized too, since there will be no harm